WIP PR for SRVKS-1077: Document volume types supported by Openshift Serverless (#67478)
openshift/openshift-docs: knative-serving/config-applications/pvcs-for-serving.adoc

= Persistent Volume Claims for Serving
:context: pvcs-for-serving

Some serverless applications need permanent data storage. You can provide data storage for Knative services by configuring different volume types. Serving supports mouthing the volume types such as `secret`, `configMap`, `projected`, and `emptyDir`. 

> You can configure persistent volume claims (PVCs) for your Knative services. The PersistentVolume types are implemented as plugins. Persistent volumes are supported but require a feature flag that must be enabled. Serverless applications also support PersistentVolume types such as `awsElasticBlockStore`, `gcePersistentDisk`, and `cephfs`.

Reviewer: I've put a comment on https://issues.redhat.com/browse/SRVKS-1077?focusedId=23518023&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-23518023 with some details around that. We should probably add more details and suggest using OpenShift Data Foundation.

Reviewer: I've just noticed that we already mention `ocs-storagecluster-cephfs`.
It would be good explain that it's coming from OpenShift Data Foundation as a tested solution: https://docs.openshift.com/container-platform/4.14/storage/persistent_storage/persistent-storage-ocs.html

Reviewer: Just to summarize the previous comments and discussion on slack:
* We should remove the sentence "Serverless applications also support PersistentVolume types such as `awsElasticBlockStore`, `gcePersistentDisk`, and `cephfs`." as these were not tested downstream.
* We should explaing that the `ocs-storagecluster-cephfs` storage class being used in the example later in this document is coming from OpenShift Data Foundation and provide a link to docs.

> Some serverless applications need permanent data storage. You can provide data storage for Knative services by configuring different volume types. Serving supports mounting the volume types such as `secret`, `configMap`, `projected`, and `emptyDir`. 
> 
> You can configure persistent volume claims (PVCs) for your Knative services. The PersistentVolume types are implemented as plugins. In order to determine if there are any persistent volume types available, you can check the available or installed storage classes in your cluster. Persistent volumes are supported but require a feature flag that must be enabled. 
> 
> [WARNING]
> ====
> The mounting of large volumes may result in a significant delay in the application's start-up time.
> ====

Reviewer 2: ```suggestion
Some serverless applications require permanent data storage. You can provide data storage for Knative services by configuring different volume types. Serving supports the volume types such as `secret`, `configMap`, `projected`, and `emptyDir`.

You can configure persistent volume claims (PVCs) for your Knative services. The PersistentVolume types are implemented as plugins. To determine if persistent volume types are available, you can check the available or installed storage classes in your cluster. Persistent volumes are supported, but require a feature flag to be enabled.

[WARNING]
====
The mounting of large volumes can lead to a considerable delay in the start time of the application.
====
```

Author: Done

= Persistent Volume Claims for Serving
:context: pvcs-for-serving

Some serverless applications need permanent data storage. You can provide data storage for Knative services by configuring different volume types such as `secret`, `configMap`, `projected`, and `emptyDir`. 

> To achieve this, you can configure persistent volume claims (PVCs) for your Knative services. The PersistentVolume types are implemented as plugins. Serverless applications also support PersistentVolume types such as `awsElasticBlockStore`, `gcePersistentDisk`, and `cephfs`.

Reviewer 3: PVCs are just another type of volume source, so let's follow the way we phrased it upstream. We recently merged the new [storage documentation](https://knative.dev/development/serving/services/storage/).

> You can provide data storage for Knative Services by configuring different volumes types. Serving supports mounting the [volume types](https://kubernetes.io/docs/concepts/storage/volumes): emptyDir, secret, configMap and projected. [PersistentVolumes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/) are supported but require a [feature flag](https://knative.dev/development/serving/configuration/feature-flags/) to be enabled.

QE needs to verify this:

> `awsElasticBlockStore`, `gcePersistentDisk`, and `cephfs`

Author: Sure! I will try to include this.

Author: @skonto I have a doubt. Our Section title is **Persistent Volume Claims for Serving** so we need to mention about PVCs as well, right? 

Also, we can not include the upstream doc links in the downstream documents. For example, [PersistentVolumes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/) or [volume types](https://kubernetes.io/docs/concepts/storage/volumes/)
